feat: Pure Python audio chat app with Multimodal Live API

freddyaboulton opened the pull request:
# Description

The `multimodal-live-app` requires knowledge of html and js. 

Adding a pure-python webapp built with Gradio that supports audio input/output streaming will help developers get started building with the multimodal api without leaving their preferred language.

- [ ] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [ ] You are listed as the author in your notebook or README file.
  - [ ] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [ ] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [ ] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [ ] Appropriate docs were updated (if necessary)

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,253 @@
+            output_frame_size=self.output_frame_size,
+        )
+
+    def _initialize_websocket(self) -> None:
+        assert self.config, "Config not set"
+        try:
+            self.ws = websockets.sync.client.connect(self.config.ws_url, timeout=30)
+            initial_request = {
+                "setup": {
+                    "model": self.config.model,
+                }
+            }
+            self.ws.send(json.dumps(initial_request))
+            setup_response = json.loads(self.ws.recv())
+            print(f"Setup response: {setup_response}")
+        except websockets.exceptions.WebSocketException as e:
+            print(f"WebSocket connection failed: {str(e)}")
+            self.ws = None
```
Can you try using the `genai.live.start_stream` method instead of raw websockets calls?

https://googleapis.github.io/python-genai/genai.html#module-genai.live

holtskinner reviewed (changes requested):
Please also resolve spelling errors and respond to Gemini review comments. Deferring to Zack/Katie for further review

freddyaboulton commented:
Thank you @holtskinner  ! I will address the Gemini comments today/tomorrow. I think there is something wrong with the spellcheck as `WebRTC` is a valid term. How can I configure it to ignore that phrase?

freddyaboulton commented:
Hi @holtskinner @katiemn @ZackAkil - I switched the demo to use `start_stream`! However, I ran into a couple of issues I had to solve manually in my locally installed version of the `genai` package. See here: https://github.com/googleapis/python-genai/issues/35

katiemn commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,128 @@
+        client = genai.Client(api_key=api_key)
+        config = {"response_modalities": ["AUDIO"]}
+        async with client.aio.live.connect(
+            model="models/gemini-2.0-flash-exp", config=config
```
It looks like this demo connects to Gemini via Google AI Studio. This repository is for demos and code samples for Vertex AI. Please check out this repo to contribute Google AI Studio samples, https://github.com/google-gemini/cookbook/

holtskinner replied:
Alternatively, if you can switch this demo to use Vertex AI instead, we're happy to host it in this repo.

freddyaboulton replied:
Hi @holtskinner @katiemn - just switched over to use Vertex AI! 

https://github.com/user-attachments/assets/727b7ef6-7072-4239-a73e-c90834e3f599

holtskinner commented:
@freddyaboulton Thanks for making the adjustments! Can you please resolve the linter errors? https://github.com/GoogleCloudPlatform/generative-ai/actions/runs/12584981682/job/35075770729?pr=1551

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,134 @@
+        self.quit.set()
+
+
+css = """
+#api-form {
+    width: 80%;
+    margin: auto;
+}
+"""
```
Can this be moved to a separate file and loaded in?

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,134 @@
+import numpy as np
+
+
+def encode_audio(data: np.ndarray, sample_rate: int) -> str:
```
`sample_rate` doesn't seem to be used. Should it be input to one of the function calls?

freddyaboulton replied:
Oversight on my part, it's not needed in the function signature.

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,134 @@
+            output_frame_size=self.output_frame_size,
+        )
+
+    async def stream(self):
+        while not self.quit.is_set():
+            audio = await self.input_queue.get()
+            yield audio
```
Can you add a return value?

freddyaboulton commented:
Hi @holtskinner ! Thanks for the patience over the holidays. I've updated the demo based on your suggestions and fixed the lint!

holtskinner commented on gemini/gradio-voice/app.py:
One more change. Can you move the `gradio-voice` directory inside the `gemini/multimodal-live-api/` directory?

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,140 @@
+                label="Project ID",
+                placeholder="Enter your Google Cloud Project ID",
+            )
+            location_ = gr.Textbox(
+                label="Location",
+                placeholder="Enter the location of your project, e.g. us-central1",
+            )
```
nit: Can you make a dropdown list for this?

https://cloud.google.com/vertex-ai/generative-ai/docs/learn/locations#genai-locations

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,140 @@
+    gr.HTML(header)
+    with gr.Group(visible=True, elem_id="api-form") as api_key_row:
+        with gr.Row():
+            project_id_ = gr.Textbox(
```
Can you add an option to select the voice from a list?

holtskinner commented on gemini/gradio-voice/app.py:
```diff
@@ -0,0 +1,168 @@
+        self, project_id: str, location: str, voice_name: str = "Puck"
+    ) -> AsyncGenerator[bytes, None]:
+        """Connect to to genai server and start the stream"""
+        client = genai.Client(vertexai=True, project=project_id, location=location)
```
Since you made `client` a field on the `GeminiHandler` class, should you assign this to `self.client`?

freddyaboulton replied:
I'll remove the field since it's only used in this method. Thanks, good catch!

freddyaboulton commented:
Hi @holtskinner ! Thanks for adding the dropdowns for region and voice! I moved the directory and added a README as well as addressing the other comments.

holtskinner commented on gemini/multimodal-live-api/gradio-voice/README.md:
```diff
@@ -0,0 +1,110 @@
+
+The main dependencies are:
+
+- gradio_webrtc>=0.0.27,<1.0
+- librosa
+- google-genai==0.3.0
```
```suggestion
- `gradio_webrtc>=0.0.27,<1.0`
- `librosa`
- `google-genai==0.3.0`
```
